crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2576,7 +2609,53 @@ impl<'db> TypeInferenceBuilder<'db> {
                         Type::Unknown
                     })
             }
-            _ => Type::Todo,
+            (value_ty, slice_ty) => {
+                // Resolve the value to its class.
+                let value_meta_ty = value_ty.to_meta_type(self.db);
+
+                // If the class defines `__getitem__`, return its return type.
+                //
+                // See: https://docs.python.org/3/reference/datamodel.html#class-getitem-versus-getitem
+                let dunder_getitem_method = value_meta_ty.member(self.db, "__getitem__");
+                if !dunder_getitem_method.is_unbound() {
+                    // TODO validate that the `__getitem__` argument adheres to the signature
+                    let CallOutcome::Callable { return_ty } =
+                        dunder_getitem_method.call(self.db, &[slice_ty])
+                    else {
+                        self.dunder_not_callable_diagnostic(
+                            (&**value).into(),
+                            value_ty,
+                            "__getitem__",
+                        );
+                        return Type::Unknown;
+                    };
+                    return return_ty;
+                }
+
+                // Otherwise, if the value is itself a class and defines `__class_getitem__`,
+                // return its return type.
+                if matches!(value_ty, Type::Class(_)) {
+                    let dunder_class_getitem_method = value_ty.member(self.db, "__class_getitem__");
+                    if !dunder_class_getitem_method.is_unbound() {
+                        // TODO validate that the `__class_getitem__` argument adheres to the signature
+                        let CallOutcome::Callable { return_ty } =
+                            dunder_class_getitem_method.call(self.db, &[slice_ty])
+                        else {
+                            self.dunder_not_callable_diagnostic(
```

Comment:
Same comment as above about using `CallOutcome::unwrap_with_diagnostic` instead of directly matching on the `CallOutcome`.